keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,392 @@
+
+Our feature does not introduce new values or new fields. It enables the usage of an existing value in existing fields, but with a different logic. I.e. if someone creates a Service with mixed protocol setup and then rollbacks the API server to a version that does not implement this feature the clients will still get the Service with mixed protocols when they read that via the rollback'd API. If the client (CPI implementation) has been rollbacked, too, then the client may receive such a Service setup that it does not support.
+
+- Alibaba: no risk. The current CPI and LB already supports the mixed protocols in the same Service definition. If this feature is enabled in an API server and then the API server rollback is executed the CPI can still handle the Services with mixed protocol sets.
+- AWS: no risk. The current CPI accepts Services with TCP protocol only, i.e. after a K8s upgrade a user still cannot use this feature. Consequently, a rollback in the K8s version does not introduce any issues.
+- Azure: no risk. The current CPI and LB already supports the mixed protocols in the same Service definition. The situation is the same as with the Alibaba CPI.
+- GCE: currently the GCE CPI assumes that a Service definition contains a single protocol value, as it assumes that the Service Controller already rejected Services with mixed protocols. While the Service Controller really did so a while ago, it does not do this anymore. It means a risk.
+- IBM Cloud VPC: no risk. The same situation like in the case of AWS.
+- IBM Cloud Classic: no risk. The CPI and NLB already supports TCP and UDP in the same Service definition. The same situation like in the case of Alibaba.
+- OpenStack: no risk. The CPI and NLB already supports TCP and UDP in the same Service definition. The same situation like in the case of Alibaba.
+- Oracle: no risk. The CPI and LB already supports mixed protocols. The same situation like in the case of Alibaba.
+- Tencent Cloud: no risk. The CPI and LB already supports mixed protocols. The same situation like in the case of Alibaba.
+
+
+The other risk is in the introduction of this feature without an option control mechanism, i.e. as a general change in Service handling. In that case there is the question whether this feature should be a part of the conformance test set, because it can affect the conformance of cloud providers.
+
+A possible mitigation is to put the feature behind option control. 
```

Comment:
If we allow existing services to have this condition in 1.18 (e.g. implementing it and not rejecting updates to objects that already are in this state) but not allow new services (e.g. rejecting a create or an update that enters this state), then we can allow create in 1.19, knowing that rollback to 1.18 is safe.

I think?  Am I missing anything?

Reply:
Seems right to me -- we followed the same approach when we added finalizers for Service LBs. The only thing to consider is the code path won't be vetted as much if we don't allow new services to use it at all, at the very least there should be an opt-in for mixed protocols so we can verify it works on v1.18 but we need to make sure users know the consequence of enabling (i.e. no rollback). Maybe we do need a feature gate?

Reply:
yeah, these changes are always nuanced.  I *think* the right thing is to allow existing uses and allow new uses IFF gate is enabled.  That puts anyone who uses it under the gate at rollback risk.

I have asked @liggitt for a consult, since he is involved in all of these :)

Reply:
> I think the right thing is to allow existing uses and allow new uses IFF gate is enabled.

That's correct. The essential thing that default configurations (e.g. no opt into alpha feature gate) of version N does not allow new uses (via creation or update of an object that doesn't currently use the relaxed validation) that would be considered invalid by version N-1.

> That puts anyone who uses it under the gate at rollback risk.

That is why we put relaxed validation and new fields in GA objects under an alpha gate for one release, and limit API server rollback/skew to a single version.

Reply:
Strictly speaking - it doesn't HAVE to be "alpha", it could just be off-by-default beta (not much difference practically).

Reply:
Fixed. Let's see if it is better now. Please check.